File: consys-core/src/main/scala/de/tuda/stg/consys/core/store/cassandra/CassandraTransactionContextBinding.scala
package de.tuda.stg.consys.core.store.cassandra

import de.tuda.stg.consys.core.store.TransactionContext

import scala.reflect.ClassTag
import scala.reflect.runtime.universe.TypeTag


/**
 * Created on 14.01.20.
 *
 * @author Mirko Köhler
 */
trait CassandraTransactionContextBinding extends TransactionContext {

	override type StoreType = CassandraStore

	override private[store] def replicateRaw[T <: StoreType#ObjType : ClassTag](addr : StoreType#Addr, obj : T, level : StoreType#Level) : StoreType#RawType[T] =
		level.toProtocol(store).writeRaw[T](addr, obj, this.asInstanceOf[StoreType#TxContext] /* TODO: Is there a better way to get a transaction context? */)

	override private[store] def lookupRaw[T <: StoreType#ObjType : ClassTag](addr : StoreType#Addr, level : StoreType#Level) : StoreType#RawType[T] =
		level.toProtocol(store).readRaw[T](addr, this.asInstanceOf[StoreType#TxContext] /* TODO: Is there a better way to get a transaction context? */)

}

File: consys-japi/src/main/java/de/tuda/stg/consys/japi/binding/cassandra/CassandraTransactionContextBinding.java
package de.tuda.stg.consys.japi.binding.cassandra;

import de.tuda.stg.consys.core.store.ConsistencyLevel;
import de.tuda.stg.consys.core.store.cassandra.CassandraRef;
import de.tuda.stg.consys.core.store.cassandra.CassandraStore;
import de.tuda.stg.consys.core.store.cassandra.CassandraTransactionContext;
import de.tuda.stg.consys.japi.Ref;
import de.tuda.stg.consys.japi.TransactionContext;

import java.io.Serializable;

public class CassandraTransactionContextBinding implements TransactionContext<String, Serializable, ConsistencyLevel<CassandraStore>> {
    private final CassandraTransactionContext ctx;

    CassandraTransactionContextBinding(CassandraTransactionContext ctx) {
        this.ctx = ctx;
    }

    @Override
    public <T extends Serializable> Ref<T> replicate(String s, ConsistencyLevel<CassandraStore> level, Class<T> clazz, Object... constructorArgs) {
        CassandraRef<T> handler = (CassandraRef<T>) ctx.replicate(s, level, clazz, constructorArgs);
        return new CassandraRefBinding<>(handler);
    }

    @Override
    public <T extends Serializable> Ref<T> lookup(String s, ConsistencyLevel<CassandraStore> level, Class<T> clazz) {
        CassandraRef<T> handler = (CassandraRef<T>) ctx.lookup(s, level, clazz);
        return new CassandraRefBinding<>(handler);
    }
}

